src/nupic/algorithms/TemporalMemory.cpp

```diff
@@ -144,415 +143,344 @@ void TemporalMemory::initialize(
+  Iterator begin()
   {
-    bool found;
-    tie(found, segment, numActiveSynapses) =
-      bestMatchingSegment(cell, activeCells, _connections);
-
-    if (found && numActiveSynapses > maxSynapses)
-    {
-      maxSynapses = numActiveSynapses;
-      foundCell = true;
-      bestCell = cell;
-      foundSegment = true;
-      bestSegment = segment;
-    }
+    return Iterator(activeColumns_.begin(),
+                    activeColumns_.end(),
+                    activeSegments_.begin(),
+                    activeSegments_.end(),
+                    matchingSegments_.begin(),
+                    matchingSegments_.end(),
+                    cellsPerColumn_);
   }
 
-  if (!foundCell)
+  Iterator end()
   {
-    bestCell = leastUsedCell(cells, _connections);
-    foundCell = true;
-    foundSegment = false;
+    return Iterator(activeColumns_.end(),
+                    activeColumns_.end(),
+                    activeSegments_.end(),
+                    activeSegments_.end(),
+                    matchingSegments_.end(),
+                    matchingSegments_.end(),
+                    cellsPerColumn_);
   }
 
-  return make_tuple(foundCell, bestCell, foundSegment, bestSegment);
-}
-
-tuple<bool, Segment, Int>
-TemporalMemory::bestMatchingSegment(
-  Cell& cell,
-  set<Cell>& activeCells,
-  Connections& _connections)
+private:
+  const vector<UInt>& activeColumns_;
+  const UInt cellsPerColumn_;
+  const vector<SegmentOverlap>& activeSegments_;
+  const vector<SegmentOverlap>& matchingSegments_;
+};
+
+void TemporalMemory::compute(
+  UInt activeColumnsSize,
+  const UInt activeColumnsUnsorted[],
+  bool learn)
 {
-  Int maxSynapses = minThreshold_;
-  Int bestNumActiveSynapses = 0;
-  Segment bestSegment;
-  bool found = false;
+  const vector<Cell> prevActiveCells = activeCells;
+  const vector<Cell> prevWinnerCells = winnerCells;
 
-  for (Segment segment : _connections.segmentsForCell(cell))
-  {
-    Int numActiveSynapses = 0;
+  vector<UInt> activeColumns(activeColumnsUnsorted,
+                             activeColumnsUnsorted + activeColumnsSize);
+  std::sort(activeColumns.begin(), activeColumns.end());
 
-    for (auto synapse : _connections.synapsesForSegment(segment))
-    {
-      SynapseData synapseData = _connections.dataForSynapse(synapse);
+  activeCells.clear();
+  winnerCells.clear();
 
-      if (synapseData.permanence > 0 &&
-          activeCells.find(synapseData.presynapticCell) != activeCells.end())
+  for (const ExcitedColumnData& excitedColumn : ExcitedColumns(activeColumns,
+                                                               activeSegments,
+                                                               matchingSegments,
+                                                               cellsPerColumn_))
+  {
+    if (excitedColumn.isActiveColumn)
+    {
+      if (excitedColumn.activeSegmentsBegin != excitedColumn.activeSegmentsEnd)
+      {
+        // Predicted active column.
```

Comment:
I'd break each of the three cases out into pure functions. That will make this function easier to read.